added CRC64/NVME as a valid checksum type (main)

JustinKyleJames opened the pull request:
Companion PR: https://github.com/irods/irods_resource_plugin_s3/pull/2270

This is to replace the pull request for #8698.

I ran the full suite of tests again for these commits.

---

From the original PR.

1. Added hash option for crc64nvme.
2. Added RESOURCE_OP_READ_CHECKSUM resource plugin operation.
3. Called the above operation before reading file to calculate checksum.  If it succeeds use the checksum returned from it.

korydraughn commented on lib/hasher/include/irods/CRC64NVMEStrategy.hpp:
```diff
@@ -0,0 +1,32 @@
+namespace irods
+{
+    extern const std::string CRC64NVME_NAME;
+    class CRC64NVMEStrategy : public HashStrategy
```
Let's add an empty line between these lines 11 and 12.

korydraughn commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,81 @@
+
+        // Create the CRC calculator object
+        boost::crc_basic<crc_bits>* context = new boost::crc_basic<crc_bits>(
+            crc_polynomial, crc_initial_remainder, crc_final_xor, crc_reflect_input, crc_reflect_output);
```
Is there a reason why the context is heap allocated?

JustinKyleJames replied:
This context needs to survive longer than this function call.

This is similar to the others.
https://github.com/irods/irods/blob/b8781c11aef62b61b3ff3e29d17b52bb105fe372/lib/hasher/src/SHA256Strategy.cpp#L30
https://github.com/irods/irods/blob/b8781c11aef62b61b3ff3e29d17b52bb105fe372/lib/hasher/src/SHA256Strategy.cpp#L86

korydraughn replied:
Sure. If the `boost:crc_basic<crc_bits>` type supports copy/move semantics, then you can avoid the heap allocation. That would also remove the need for anyone to protect against memory leaks.

With that said, code which relies on these hashers might assume data generated by them are heap allocated. If that's true, then we should continue with the existing approach.

I'm fine with your solution overall. Feel free to mark this as resolved.

korydraughn commented on server/core/include/irods/irods_resource_constants.hpp:
```diff
@@ -72,6 +72,7 @@ namespace irods
     extern const std::string RESOURCE_OP_RESOLVE_RESC_HIER;
     extern const std::string RESOURCE_OP_REBALANCE;
     extern const std::string RESOURCE_OP_NOTIFY;
+    extern const std::string RESOURCE_OP_READ_CHECKSUM;
```
Maybe this should be `RESOURCE_OP_READ_CHECKSUM_FROM_STORAGE_DEVICE`?

That makes it more specific and gives us more wiggle room for future checksum operations to be added to the resource plugins.

JustinKyleJames replied:
Done - also in S3 plugin.

korydraughn commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,81 @@
+#include <sstream>
+
+#include <openssl/err.h>
+#include <openssl/evp.h>
```
Are the OpenSSL headers actually used?

JustinKyleJames replied:
No.  Removed.

korydraughn commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,81 @@
+
+#include <fmt/format.h>
+
+#include <boost/algorithm/string/predicate.hpp>
```
This doesn't look like it's needed anymore.

korydraughn commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        int len = strlen(CRC64NVME_CHKSUM_PREFIX);
+        unsigned long out_len = CHKSUM_LEN - len;
```
Looks like these lines can be combined since `len` is not used elsewhere.

korydraughn commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,81 @@
+#include <cstring>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
```
These headers do not appear to be used.

JustinKyleJames commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,81 @@
+        uint64_t checksum = context->checksum();
+
+        // Convert uint64_t to bytes (big-endian for consistency)
+        std::array<unsigned char, 8> bytes;
```
done

JustinKyleJames commented on server/api/src/rsFileChksum.cpp:
```diff
@@ -334,6 +335,33 @@ int file_checksum(RsComm* _comm,
+            __func__,
+            ret_err.result()));
+    }
+    // ----------
```
done

JustinKyleJames commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,81 @@
+
+#include "irods/rodsErrorTable.h"
+#include "irods/checksum.h"
+#include "irods/base64.hpp"
```
done

JustinKyleJames commented on lib/hasher/include/irods/CRC64NVMEStrategy.hpp:
```diff
@@ -0,0 +1,32 @@
+
+#include "irods/HashStrategy.hpp"
+
+#include <boost/any.hpp>
```
removed

korydraughn commented on server/api/src/rsFileChksum.cpp:
```diff
@@ -349,18 +348,17 @@ int file_checksum(RsComm* _comm,
             strncpy(_calculated_checksum, checksum_from_resource.c_str(), NAME_LEN);
             return 0;
         }
         log_api::trace(fmt::format(
```
The explicit call to `fmt::format` can be removed.

korydraughn replied:
Your comment states the work was done, but `fmt::format` is still in the code?

Perhaps the change wasn't pushed?

SwooshyCueb commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,71 @@
+#include <boost/crc.hpp>
+
+#include <array>
+#include <cstring>
```
```suggestion
#include <cstdint>
#include <cstring>
```

SwooshyCueb commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,71 @@
+    {
+        boost::crc_basic<crc_bits>* context = boost::any_cast<boost::crc_basic<crc_bits>*>(_context);
+
+        uint64_t checksum = context->checksum();
```
```suggestion
        std::uint64_t checksum = context->checksum();
```

SwooshyCueb commented on lib/hasher/include/irods/CRC64NVMEStrategy.hpp:
```diff
@@ -0,0 +1,32 @@
+
+#include "irods/HashStrategy.hpp"
+
+#include <string>
```
```suggestion
#include "irods/HashStrategy.hpp"
#include "irods/irods_error.hpp"

#include <boost/any.hpp>

#include <cstdlib>
#include <string>
```

JustinKyleJames replied:
done

SwooshyCueb commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,71 @@
+
+        // Convert uint64_t to bytes (big-endian for consistency)
+        std::array<unsigned char, 8> bytes;
+        for (size_t i = 0; i < bytes.size(); ++i) {
```
```suggestion
        for (std::size_t i = 0; i < bytes.size(); ++i) {
```

JustinKyleJames replied:
done

SwooshyCueb commented on server/api/src/rsFileChksum.cpp:
```diff
@@ -334,6 +335,31 @@ int file_checksum(RsComm* _comm,
+        ret_err = resc->call<const std::string*, std::string*>(
+            _comm, irods::RESOURCE_OP_READ_CHECKSUM_FROM_STORAGE_DEVICE, file_ptr, &final_scheme_str, &checksum_from_resource);
+        if (ret_err.ok()) {
+            strncpy(_calculated_checksum, checksum_from_resource.c_str(), NAME_LEN);
```
Add `<cstring>` include and
```suggestion
            std::strncpy(_calculated_checksum, checksum_from_resource.c_str(), NAME_LEN);
```

JustinKyleJames replied:
done

SwooshyCueb commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,71 @@
+#include <boost/crc.hpp>
+
+#include <array>
+#include <cstring>
```
```suggestion
#include "irods/base64.hpp"
#include "irods/checksum.h"
#include "irods/irods_error.hpp"
#include "irods/rodsErrorTable.h"

#include <boost/any.hpp>
#include <boost/crc.hpp>

#include <array>
#include <cstdint>
#include <cstring>
```

JustinKyleJames replied:
This contradicts https://github.com/irods/irods/pull/8701#discussion_r2479225015.

I added the `cstdint` include.

@korydraughn - thoughts

JustinKyleJames commented on server/api/src/rsFileChksum.cpp:
```diff
@@ -334,6 +335,33 @@ int file_checksum(RsComm* _comm,
+            "checksum read not supported by resource.  Checksum will be calculated by a full file read.  {}",
+            __func__,
+            ret_err.result()));
+    }
```
> As long as the new fileDriver call is made from this API, the redirection should be handled. Take a look at `rsFileChksum` on L42 in this file and you'll see the redirection logic there. Once it has redirected to the appropriate server, it doesn't _really_ matter how the resource operation is invoked other than convention and potentially code reuse, I guess.

Yes, I didn't get exactly what Kory was asking me to do.

JustinKyleJames replied:
done

JustinKyleJames commented on server/core/src/irods_resource_constants.cpp:
```diff
@@ -42,7 +42,7 @@ namespace irods
     const std::string RESOURCE_OP_REBALANCE( "resource_rebalance" );
     const std::string RESOURCE_OP_NOTIFY( "resource_notify" );
-    const std::string RESOURCE_OP_READ_CHECKSUM("resource_read_checksum");
+    const std::string RESOURCE_OP_READ_CHECKSUM_FROM_STORAGE_DEVICE("resource_read_checksum");
```
done

JustinKyleJames commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,71 @@
+        unsigned long out_len = CHKSUM_LEN - strlen(CRC64NVME_CHKSUM_PREFIX);
+
+        std::array<unsigned char, CHKSUM_LEN> out_buffer;
+        base64_encode(bytes.data(), bytes.size(), out_buffer.data(), &out_len);
```
done

JustinKyleJames commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,71 @@
+    {
+        constexpr uint64_t crc_polynomial = 0xAD93D23594C93659;
+        constexpr uint64_t crc_initial_remainder = 0xFFFFFFFFFFFFFFFF;
+        constexpr uint64_t crc_final_xor = 0xFFFFFFFFFFFFFFFF;
```
done

JustinKyleJames commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,81 @@
+    {
+        constexpr uint64_t crc_polynomial = 0xAD93D23594C93659;
+        constexpr uint64_t crc_initial_remainder = 0xFFFFFFFFFFFFFFFF;
+        constexpr uint64_t crc_final_xor = 0xFFFFFFFFFFFFFFFF;
```
Done - with help of chatgpt.

trel replied:
looks good

alanking commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -41,18 +42,22 @@ namespace irods
+		if (0 != rc) {
+			const auto msg = fmt::format("{}: Failed to b64 encode hash.", __func__);
+			return ERROR(rc, msg);
+		}
```
I'm guessing `delete context` is omitted here on purpose?

korydraughn commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,92 @@
+    {
+        return _chksum.starts_with(CRC64NVME_CHKSUM_PREFIX);
+    }
+}; // namespace irods
```
```suggestion
} // namespace irods
```

korydraughn commented on server/drivers/include/irods/fileDriver.hpp:
```diff
@@ -44,4 +44,5 @@ irods::error fileRegistered( rsComm_t* _comm, irods::first_class_object_ptr _obj
 irods::error fileUnregistered( rsComm_t* _comm, irods::first_class_object_ptr _object );
 irods::error fileModified( rsComm_t* _comm, irods::first_class_object_ptr _object );
 irods::error fileNotify( rsComm_t* _comm, irods::first_class_object_ptr _object , const std::string& );
+irods::error fileChecksum( rsComm_t* _comm, irods::first_class_object_ptr _object , const std::string&, std::string&);
```
Why `fileChecksum` instead of `fileReadChecksumFromStorageDevice`?

JustinKyleJames replied:
done

alanking reviewed (commented):
Had one question

korydraughn commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,92 @@
+        std::array<unsigned char, CHKSUM_LEN> out_buffer;
+        int rc = base64_encode(bytes.data(), bytes.size(), out_buffer.data(), &out_len);
+		if (0 != rc) {
+			const auto msg = fmt::format("{}: Failed to b64 encode hash.", __func__);
```
b64 -> base64

JustinKyleJames commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
Looking at `base64_encode` the only way it fails is if the buffer (size identified by `out_len`) was not large enough.

It isn't possible to call this a second time and get a different result since the following would always give the same result:

```
unsigned long out_len = CHKSUM_LEN - std::strlen(CRC64NVME_CHKSUM_PREFIX);
```

So in my eyes is makes sense to clean up the memory.  If the caller happened to try again, they are just going to get a null pointer error returned.

I'll wait to see if @SwooshyCueb has a different opinion on this.

JustinKyleJames replied:
I am resolving this.

korydraughn commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        if (nullptr == context) {
+            return ERROR(INVALID_INPUT_ARGUMENT_NULL_POINTER, "A null context was send to CRC64NVMEStrategy::update.");
```
send -> sent

korydraughn commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        if (nullptr == context) {
+            return ERROR(INVALID_INPUT_ARGUMENT_NULL_POINTER, "A null context was send to CRC64NVMEStrategy::digest.");
```
send -> sent

JustinKyleJames replied:
done

korydraughn commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,117 @@
+            delete context;
+            _context = nullptr;
+            const auto msg = fmt::format("{}: Failed to base64 encode hash.", __func__);
+            return ERROR(rc, msg);
```
These lines can be merged together.

alanking reviewed (commented):
Seems good to me. Maybe ready for squashing?

JustinKyleJames commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,117 @@
+        try {
+            context = boost::any_cast<boost::crc_basic<crc_bits>*>(_context);
+        }
+        catch (boost::bad_any_cast& e) {
```
done

JustinKyleJames commented on lib/hasher/src/CRC64NVMEStrategy.cpp:
```diff
@@ -0,0 +1,117 @@
+        try {
+            context = boost::any_cast<boost::crc_basic<crc_bits>*>(_context);
+        }
+        catch (boost::bad_any_cast& e) {
```
done

JustinKyleJames commented:
I added three test cases for CRC64NVME.  This are similar to the tests that are in the S3 resource plugin.

These are in:
- scripts/irods/test/resource_suite.py
- scripts/irods/test/test_ichksum.py

I tested these with the following but will keep off a full suite:

```
python3 scripts/run_tests.py --run_s=test_resource_types.Test_Resource_Unixfilesystem.test_iput_with_crc64nvme_lowercase_checksum_flag
python3 scripts/run_tests.py --run_s=test_resource_types.Test_Resource_Unixfilesystem.test_iput_with_crc64nvme_uppercase_checksum_flag
python3 scripts/run_tests.py --run_s=test_ichksum.Test_Ichksum.test_ichksum_with_crc64nvme
```
I also added in a `calculate_crc64_nvme()` method into lib.py.  This is also a copy of the one in s3plugin_lib.py.  I wasn't sure if I should remove the latter or keep it so I kept it for now.

Finally, the imports are not sorted in test_ichksum.py so I just added the two new ones at the bottom.

JustinKyleJames commented:
> > I wasn't sure if I should remove the latter or keep it so I kept it for now.
> 
> To me, this depends on whether the next S3 plugin release needs to be compatible with 5.0 in addition to 5.1. If so, the new function needs to be in both. If it is only going to be compatible with 5.1, then we only need the one in `lib.py`.

It has to have the CRC/NVME core code otherwise these tests will fail anyway. So I am thinking remove from s3 lib and use the one in lib.py.

korydraughn commented:
Have you tested things in a topology, with redirection?

JustinKyleJames commented on scripts/irods/test/resource_suite.py:
```diff
@@ -496,6 +497,78 @@ def test_local_iput_upper_checksum(self):
+            lib.make_arbitrary_file(file1, file1_size)
+
+            with lib.file_backed_up(config.server_config_path):
+
```
I like the empty line there but will remove.

JustinKyleJames commented on scripts/irods/test/resource_suite.py:
```diff
@@ -496,6 +497,78 @@ def test_local_iput_upper_checksum(self):
+            lib.make_arbitrary_file(file1, file1_size)
+
+            with lib.file_backed_up(config.server_config_path):
+
```
done

JustinKyleJames commented on scripts/irods/test/test_ichksum.py:
```diff
@@ -571,3 +573,39 @@ def test_ichksum_server_rejects_invalid_server_config_options_for_checksum_read_
+            lib.make_arbitrary_file(file1, file1_size)
+
+            with lib.file_backed_up(config.server_config_path):
+
```
done

JustinKyleJames commented on scripts/irods/test/resource_suite.py:
```diff
@@ -496,6 +497,78 @@ def test_local_iput_upper_checksum(self):
+            lib.remove_file_if_exists(file1)
+            IrodsController().reload_configuration()
+
+    def test_iput_with_crc64nvme_uppercase_checksum_flag(self):
```
done

korydraughn commented:
Anything else left to do for this?

korydraughn commented:
My bad on the line width comment. I was referring to the commit message, not the code. So you can revert the changes to the test or keep them if you like.

korydraughn commented:
Merging.
